### PR TITLE
Add fullsync stat caching on a configurable interval.

### PR DESCRIPTION
**Pull request by lordnull**
A rebased and squashed version of https://github.com/basho/riak_repl/pull/548 .

### CODE REVIEW

File: src/riak_repl2_fscoordinator.erl [hunk 1]
```diff
@@ -740,6 +781,52 @@ sort_partitions(In, N, Acc) ->
...
+    receive
+        refresh_stats -> ok
+    after 0 ->
+        ok
+    end,
+    StatCache#stat_cache{refresh_timer = undefined}.
+
+schedule_stat_refresh(StatCache) ->
+    StatCache1 = maybe_cancel_timer(StatCache),
+    Timer = erlang:send_after(?DEFAULT_STAT_REFRESH_INTERVAL, self(), refresh_stats),
```
Comment (andrewjstone) [inline src/riak_repl2_fscoordinator.erl]: This should use `StatCache#stat_cache.refresh_interval` instead of `?DEFAULT_STAT_REFRESH_INTERVAL`

**Comment by andrewjstone** [comment]
- [x] code inspection
- [x] eunit + eqc
- [ ] replication2 riak_test
- [x] repl_fs_stat_caching riak_test
- [x] dialyzer
- [x] xref

**Comment by andrewjstone** [comment]
Hmmm. Replication2 is failing a lot. I've run it five times and it only passed once.

Here's the failure I'm seeing: 

```
13:52:37.726 [info] waiting for fullsync count to be 2
13:52:37.726 [info] Starting fullsync on 'dev2@127.0.0.1' (current)
14:02:36.985 [error] Error in process <0.100.0> on node 'riak_test@127.0.0.1' with exit value: {{assertEqual_failed,[{module,repl_util},{line,213},{expression,"Res"},{expected,ok},{value,{fail,false}}]},[{repl_util,'-start_and_wait_until_fullsync_complete/2-fun-3-',2,[{file,"tests/repl_util.erl"},{line,213}]},{repl_util...


14:02:36.985 [warning] replication2 failed: {{assertEqual_failed,[{module,repl_util},{line,213},{expression,"Res"},{expected,ok},{value,{fail,false}}]},[{repl_util,'-start_and_wait_until_fullsync_complete/2-fun-3-',2,[{file,"tests/repl_util.erl"},{line,213}]},{repl_util,start_and_wait_until_fullsync_complete,2,[{file,"tests/repl_util.erl"},{line,213}]},{replication2,replication,3,[{file,"tests/replication2.erl"},{line,258}]},{replication2,confirm,0,[{file,"tests/replication2.erl"},{line,52}]}]}
14:02:36.985 [error]
================ replication2 failure stack trace =====================
{{assertEqual_failed,[{module,repl_util},
                      {line,213},
                      {expression,"Res"},
                      {expected,ok},
                      {value,{fail,false}}]},
 [{repl_util,'-start_and_wait_until_fullsync_complete/2-fun-3-',2,
             [{file,"tests/repl_util.erl"},{line,213}]},
  {repl_util,start_and_wait_until_fullsync_complete,2,
             [{file,"tests/repl_util.erl"},{line,213}]},
  {replication2,replication,3,[{file,"tests/replication2.erl"},{line,258}]},
  {replication2,confirm,0,[{file,"tests/replication2.erl"},{line,52}]}]}
=======================================================================

14:02:36.985 [notice] replication2 Test Run Complete

Test Results:
replication2-bitcask: fail
---------------------------------------------
1 Tests Failed
0 Tests Passed
That's 0.0% for those keeping score
```

**Comment by cmeiklejohn** [comment]
How does replication2 run without the patch?  This error appears to be just not waiting long enough for the fullsync to complete.

**Comment by andrewjstone** [comment]
I've seen it pass every time on develop lately. I'll switch and run it a few more times there to make sure. If I see the same failure I'll +1.

**Comment by cmeiklejohn** [comment]
Ah, I'm not sure that's the right solution.  It would be nice to figure out why things are so much slower.  I can try running it on EC2 if we need to.

**Comment by andrewjstone** [comment]
Yes, obviously not optimal. Replication2 just passed twice on develop for me. Something in this PR is breaking it. I can try to track it down.

**Comment by andrewjstone** [comment]
Grrrr. I just added some `lager:info` calls locally to `repl_util:start_and_wait_until_fullsync_complete/2` and changed the timing. This branch passed replication2 twice in a row. Still working on tracking it down.

**Comment by andrewjstone** [comment]
Ok it failed again. But the error is different this time:

```
17:02:37.480 [info] Fullsync on 'dev3@127.0.0.1' complete
17:02:37.480 [info] read 2000 keys
17:02:40.128 [info] Reads: []
17:02:40.131 [info] Testing master failover: stopping 'dev3@127.0.0.1'
17:02:40.131 [info] Stopping riak on 'dev3@127.0.0.1'
17:02:50.132 [info] Running: /Users/ajs/rt/riak_ee/current/dev/dev3/bin/riak stop
17:02:53.846 [info] Wait until 'dev3@127.0.0.1' is not pingable
17:02:53.848 [info] New leader is 'dev1@127.0.0.1'
17:02:53.849 [info] Writing 100 more keys to 'dev1@127.0.0.1' now that the old leader is down
17:05:54.032 [warning] 3 errors while writing: [{245,{error,timeout}},{241,{error,timeout}},{231,{error,timeout}}]
17:08:55.040 [error] Error in process <0.100.0> on node 'riak_test@127.0.0.1' with exit value: {{assertEqual_failed,[{module,replication2},{line,211},{expression,"repl_util : do_write ( ASecond , 201 , 300 , TestBucket , 2 )"},{expected,[]},{value,[{245,{error,timeout}},{241,{error,timeout}},{231,{error,timeout}}]}]},[{replication2...


17:08:55.040 [warning] replication2 failed: {{assertEqual_failed,[{module,replication2},{line,211},{expression,"repl_util : do_write ( ASecond , 201 , 300 , TestBucket , 2 )"},{expected,[]},{value,[{245,{error,timeout}},{241,{error,timeout}},{231,{error,timeout}}]}]},[{replication2,'-replication/3-fun-39-',3,[{file,"tests/replication2.erl"},{line,211}]},{replication2,replication,3,[{file,"tests/replication2.erl"},{line,211}]},{replication2,confirm,0,[{file,"tests/replication2.erl"},{line,52}]}]}
17:08:55.040 [error]
================ replication2 failure stack trace =====================
{{assertEqual_failed,[{module,replication2},
                      {line,211},
                      {expression,"repl_util : do_write ( ASecond , 201 , 300 , TestBucket , 2 )"},
                      {expected,[]},
                      {value,[{245,{error,timeout}},
                              {241,{error,timeout}},
                              {231,{error,timeout}}]}]},
 [{replication2,'-replication/3-fun-39-',3,
                [{file,"tests/replication2.erl"},{line,211}]},
  {replication2,replication,3,[{file,"tests/replication2.erl"},{line,211}]},
  {replication2,confirm,0,[{file,"tests/replication2.erl"},{line,52}]}]}
=======================================================================

17:08:55.040 [notice] replication2 Test Run Complete

Test Results:
replication2-bitcask: fail
---------------------------------------------
1 Tests Failed
0 Tests Passed
That's 0.0% for those keeping score
```

**Comment by lordnull** [comment]
I can't get develop replication2 to pass or fail with consistency; running under an ubuntu vm. So, I feel your pain and am not really any closer to tracking down the error either :/.

**Comment by andrewjstone** [comment]
This status function is being called via `status/0` and looks like it can go cross node. It should have a timeout as far as I can tell. https://github.com/basho/riak_repl/blob/d608bb7b1202d6e326a8f6547ba79e69c4f571ec/src/riak_repl2_fscoordinator.erl#L141

**Comment by andrewjstone** [comment]
Just got another 5 passes in a row of `replication2` on this branch.

**Comment by andrewjstone** [comment]
@lordnull @cmeiklejohn My current vote is to merge this, since we've been seeing periodic failures both on this branch and develop. However, I understand if that is not what's wanted. Unfortunately my debugging has turned to spinning wheels at this point. Maybe someone better at repl can handle the failures.

**Comment by cmeiklejohn** [comment]
Let me try a few runs of this against EC2 on Monday.

**Comment by cmeiklejohn** [comment]
:+1:, let buildbot take the rest of the way.  replication2 is too flappy locally.

**Pull request merged**